### ISSUE TITLE
[Enhancement] BindableObject perf improvements

### ISSUE THREAD

**Issue by andreinitescu** [xamarin/Xamarin.Forms#5015]
## Summary

[BindableObject ](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/BindableObject.cs) has several overloads of a GetValues method with the purpose to obtain in a single call the values of several bindable properties:

		[EditorBrowsable(EditorBrowsableState.Never)]
		public object[] GetValues(BindableProperty property0, BindableProperty property1)

		[EditorBrowsable(EditorBrowsableState.Never)]
		public object[] GetValues(BindableProperty property0, BindableProperty property1, BindableProperty property2)

		[EditorBrowsable(EditorBrowsableState.Never)]
		internal object[] GetValues(params BindableProperty[] properties) 

Each implementation of GetValues has same pattern, it allocates an array of objects on the **heap** and returns it, like the following:

		internal object[] GetValues(...)
		{
			var values = new object[length];
                        
			// Fill values ...

			return values;
		}

## Changes

I was wondering if allocation on heap can't be avoided by simple refactoring so that instead of calling GeValues call BindableObject.GetValue directly instead. Shouldn't GetValues work exactly like GetValue? (even better why not use the actual getter whenever possible?)

For example, in FontElement which is used by all the controls supporting text, instead of:

		static void OnFontSizeChanged(BindableObject bindable, object oldValue, object newValue)
		{
			...
			var values = bindable.GetValues(FontFamilyProperty, FontAttributesProperty);
			var fontSize = (double)newValue;
			var fontAttributes = (FontAttributes)values[1];
			var fontFamily = (string)values[0];

			if (fontFamily != null)
				bindable.SetValue(FontProperty, Font.OfSize(fontFamily, fontSize).WithAttributes(fontAttributes));
			else
				bindable.SetValue(FontProperty, Font.SystemFontOfSize(fontSize, fontAttributes));
			...
		}

replace with:

		static void OnFontSizeChanged(BindableObject bindable, object oldValue, object newValue)
		{
			...
			var fontFamily = (string)bindable.GetValue(FontFamilyProperty);
			var fontAttributes = (FontAttributes)(string)bindable.GetValue(FontAttributesProperty);
			var fontSize = (double)newValue;

			if (fontFamily != null)
				bindable.SetValue(FontProperty, Font.OfSize(fontFamily, fontSize).WithAttributes(fontAttributes));
			else
				bindable.SetValue(FontProperty, Font.SystemFontOfSize(fontSize, fontAttributes));
			...
		}

Another example is in FlexLayout implementation which for every item it allocates on the heap an array of 9 objects just to return values:

        public class FlexLayout : Layout<View>
	{
                ...
		void InitItemProperties(View view, Flex.Item item)
		{
			var values = view.GetValues(OrderProperty,
										GrowProperty,
										ShrinkProperty,
										BasisProperty,
										AlignSelfProperty,
										MarginProperty,
										WidthRequestProperty,
										HeightRequestProperty,
										IsVisibleProperty);

			item.Order = (int)values[0];
			item.Grow = (float)values[1];
			item.Shrink = (float)values[2];
			item.Basis = ((FlexBasis)values[3]).ToFlexBasis();
			item.AlignSelf = (Flex.AlignSelf)(FlexAlignSelf)values[4];
			item.MarginLeft = (float)((Thickness)values[5]).Left;
			item.MarginTop = (float)((Thickness)values[5]).Top;
			item.MarginRight = (float)((Thickness)values[5]).Right;
			item.MarginBottom = (float)((Thickness)values[5]).Bottom;
			item.Width = (double)values[6] < 0 ? float.NaN : (float)(double)values[6];
			item.Height = (double)values[7] < 0 ? float.NaN : (float)(double)values[7];
			item.IsVisible = (bool)values[8];

I also noticed that sometimes getting all values at once is not actually necessary. One example is in iOS LabelRenderer, where getting the value of the Label.Text property is only necessary if Label.FormattedText is null, please see [UpdateText()](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs#L342)

**Comment by andreinitescu** [comment]
I could help with this if it's something team thinks it makes sense

**Comment by andreinitescu** [comment]
I see the tag is "Needs Specification"
What kind of specification is it needed?  It should be just a matter of replacing some calls to avoid allocation. This should be quite straightforward to implement, and it improves performance.

**Comment by StephaneDelcroix** [comment]
In our experience, and our measurements, at the time we wrote that we figured that the paying the allocation cost was worth the perf boost of getting multiple properties at one

UPDATE:
I looked at the code, and the loop optimization is gone! GONE!

so, yes, at this time, there's no benefit of retrieving multiple properties at the same time. I'll try to look deeper into this, look at the original code, and try to decide what to do.
